File: android/speech/src/test/kotlin/io/rynbridge/speech/SpeechModuleTest.kt
package io.rynbridge.speech

import io.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class SpeechModuleTest {

    @Test
    fun `startRecognition returns sessionId`() = runTest {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        val handler = module.actions["startRecognition"]!!

        val result = handler(mapOf("language" to BridgeValue.string("en-US")))
        assertEquals("session-1", result["sessionId"]?.stringValue)
        assertEquals("en-US", provider.lastLanguage)
    }

    @Test
    fun `stopRecognition returns transcript`() = runTest {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        val handler = module.actions["stopRecognition"]!!

        val result = handler(mapOf("sessionId" to BridgeValue.string("session-1")))
        assertEquals("Hello world", result["transcript"]?.stringValue)
    }

    @Test
    fun `speak with options`() = runTest {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        val handler = module.actions["speak"]!!

        val result = handler(mapOf(
            "text" to BridgeValue.string("Say this"),
            "language" to BridgeValue.string("en-US"),
            "rate" to BridgeValue.double(1.5),
            "pitch" to BridgeValue.double(1.0),
            "voiceId" to BridgeValue.string("voice-1")
        ))
        assertTrue(result.isEmpty())
        assertNotNull(provider.lastSpeakOptions)
        assertEquals("Say this", provider.lastSpeakOptions!!.text)
        assertEquals("en-US", provider.lastSpeakOptions!!.language)
        assertEquals(1.5, provider.lastSpeakOptions!!.rate)
        assertEquals(1.0, provider.lastSpeakOptions!!.pitch)
        assertEquals("voice-1", provider.lastSpeakOptions!!.voiceId)
    }

    @Test
    fun `stopSpeaking calls provider`() = runTest {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        val handler = module.actions["stopSpeaking"]!!

        val result = handler(emptyMap())
        assertTrue(result.isEmpty())
        assertTrue(provider.stopSpeakingCalled)
    }

    @Test
    fun `getVoices returns voice list`() = runTest {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        val handler = module.actions["getVoices"]!!

        val result = handler(emptyMap())
        val voices = result["voices"]?.arrayValue
        assertNotNull(voices)
        assertEquals(1, voices!!.size)
        assertEquals("voice-1", voices[0].dictionaryValue?.get("id")?.stringValue)
        assertEquals("Samantha", voices[0].dictionaryValue?.get("name")?.stringValue)
    }

    @Test
    fun `requestPermission returns granted`() = runTest {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        val handler = module.actions["requestPermission"]!!

        val result = handler(emptyMap())
        assertEquals(true, result["granted"]?.boolValue)
    }

    @Test
    fun `getPermissionStatus returns status`() = runTest {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        val handler = module.actions["getPermissionStatus"]!!

        val result = handler(emptyMap())
        assertEquals("granted", result["status"]?.stringValue)
    }

    @Test
    fun `module name and version`() {
        val provider = MockSpeechProvider()
        val module = SpeechModule(provider)
        assertEquals("speech", module.name)
        assertEquals("0.1.0", module.version)
    }
}

private class MockSpeechProvider : SpeechProvider {
    var lastLanguage: String? = null
    var lastSpeakOptions: SpeakOptions? = null
    var stopSpeakingCalled = false

    override suspend fun startRecognition(language: String?): StartRecognitionResult {
        lastLanguage = language
        return StartRecognitionResult(sessionId = "session-1")
    }

    override suspend fun stopRecognition(sessionId: String): StopRecognitionResult =
        StopRecognitionResult(transcript = "Hello world")

    override suspend fun speak(options: SpeakOptions) {
        lastSpeakOptions = options
    }

    override fun stopSpeaking() {
        stopSpeakingCalled = true
    }

    override suspend fun getVoices(): GetVoicesResult =
        GetVoicesResult(voices = listOf(Voice(id = "voice-1", name = "Samantha", language = "en-US")))

    override suspend fun requestPermission(): PermissionResult =
        PermissionResult(granted = true)

    override suspend fun getPermissionStatus(): PermissionStatusResult =
        PermissionStatusResult(status = "granted")
}

File: packages/speech/src/__tests__/SpeechModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { SpeechModule } from '../SpeechModule.js';

describe('SpeechModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let speech: SpeechModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    speech = new SpeechModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  function simulateNativeEvent(module: string, action: string, payload: Record<string, unknown>) {
    const event = JSON.stringify({
      id: 'event-' + Math.random().toString(36).slice(2),
      module,
      action,
      payload,
      version: '1.0.0',
    });
    transport.simulateIncoming(event);
  }

  describe('startRecognition', () => {
    it('sends correct module and action without payload', async () => {
      const promise = speech.startRecognition();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('speech');
      expect(sent.action).toBe('startRecognition');
      expect(sent.payload).toEqual({});

      respondSuccess({ sessionId: 'session-123' });

      const result = await promise;
      expect(result).toEqual({ sessionId: 'session-123' });
    });

    it('sends language option', async () => {
      const promise = speech.startRecognition({ language: 'ko-KR' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.payload).toEqual({ language: 'ko-KR' });

      respondSuccess({ sessionId: 'session-456' });

      const result = await promise;
      expect(result).toEqual({ sessionId: 'session-456' });
    });
  });

  describe('stopRecognition', () => {
    it('sends sessionId and returns transcript', async () => {
      const promise = speech.stopRecognition({ sessionId: 'session-123' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('speech');
      expect(sent.action).toBe('stopRecognition');
      expect(sent.payload).toEqual({ sessionId: 'session-123' });

      respondSuccess({ transcript: 'Hello world' });

      const result = await promise;
      expect(result).toEqual({ transcript: 'Hello world' });
    });
  });

  describe('onRecognitionResult', () => {
    it('subscribes to recognition result events and returns unsubscribe function', () => {
      const received: unknown[] = [];
      const unsubscribe = speech.onRecognitionResult((data) => {
        received.push(data);
      });

      simulateNativeEvent('speech', 'recognitionResult', { transcript: 'Hello', isFinal: false });

      expect(received).toHaveLength(1);
      expect(received[0]).toEqual({ transcript: 'Hello', isFinal: false });

      simulateNativeEvent('speech', 'recognitionResult', { transcript: 'Hello world', isFinal: true });

      expect(received).toHaveLength(2);
      expect(received[1]).toEqual({ transcript: 'Hello world', isFinal: true });

      unsubscribe();
      simulateNativeEvent('speech', 'recognitionResult', { transcript: 'After unsubscribe', isFinal: false });

      expect(received).toHaveLength(2);
    });
  });

  describe('speak', () => {
    it('sends text to speak', async () => {
      const promise = speech.speak({ text: 'Hello world' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('speech');
      expect(sent.action).toBe('speak');
      expect(sent.payload).toEqual({ text: 'Hello world' });

      respondSuccess();

      await promise;
    });

    it('sends all options', async () => {
      const promise = speech.speak({
        text: 'Test',
        language: 'en-US',
        rate: 1.5,
        pitch: 0.8,
        voiceId: 'voice-1',
      });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.payload).toEqual({
        text: 'Test',
        language: 'en-US',
        rate: 1.5,
        pitch: 0.8,
        voiceId: 'voice-1',
      });

      respondSuccess();

      await promise;
    });
  });

  describe('stopSpeaking', () => {
    it('sends fire-and-forget', () => {
      speech.stopSpeaking();

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('speech');
      expect(sent.action).toBe('stopSpeaking');
      expect(sent.payload).toEqual({});
    });
  });

  describe('getVoices', () => {
    it('returns available voices', async () => {
      const promise = speech.getVoices();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('speech');
      expect(sent.action).toBe('getVoices');

      respondSuccess({
        voices: [
          { id: 'voice-1', name: 'Siri Female', language: 'en-US' },
          { id: 'voice-2', name: 'Siri Male', language: 'ko-KR' },
        ],
      });

      const result = await promise;
      expect(result).toEqual({
        voices: [
          { id: 'voice-1', name: 'Siri Female', language: 'en-US' },
          { id: 'voice-2', name: 'Siri Male', language: 'ko-KR' },
        ],
      });
    });
  });

  describe('requestPermission', () => {
    it('returns permission result', async () => {
      const promise = speech.requestPermission();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('speech');
      expect(sent.action).toBe('requestPermission');

      respondSuccess({ granted: true });

      const result = await promise;
      expect(result).toEqual({ granted: true });
    });
  });

  describe('getPermissionStatus', () => {
    it('returns permission status', async () => {
      const promise = speech.getPermissionStatus();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('speech');
      expect(sent.action).toBe('getPermissionStatus');

      respondSuccess({ status: 'granted' });

      const result = await promise;
      expect(result).toEqual({ status: 'granted' });
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = speech.startRecognition();

      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Speech recognition unavailable' },
      };
      transport.simulateIncoming(JSON.stringify(response));

      await expect(promise).rejects.toThrow('Speech recognition unavailable');
    });
  });
});
